src/storage/src/source/healthcheck.rs

```diff
@@ -0,0 +1,809 @@
+        ];
+
+        for test_case in test_cases {
+            run_test(test_case)
+        }
+
+        fn run_test(test_case: (SourceStatus, Vec<SourceStatus>, bool)) {
+            let (from_status, to_status, allowed) = test_case;
+            for status in to_status {
+                assert_eq!(allowed, from_status.can_transition(&status))
+            }
+        }
+    }
+
+    // Auxiliary functions
+    static PERSIST_CACHE: Lazy<Arc<Mutex<PersistClientCache>>> = Lazy::new(|| {
+        Arc::new(Mutex::new(PersistClientCache::new(
+            PersistConfig::new(SYSTEM_TIME.clone()),
+            &MetricsRegistry::new(),
+        )))
+    });
+
+    static PERSIST_LOCATION: Lazy<PersistLocation> = Lazy::new(|| PersistLocation {
+        blob_uri: "mem://".to_owned(),
+        consensus_uri: "mem://".to_owned(),
+    });
+
+    async fn new_healthchecker(
+        status_shard_id: ShardId,
+        source_name: String,
+        source_id: GlobalId,
+        source_type: &'static str,
+        upstream_name: Option<String>,
+        worker_id: usize,
+        worker_count: usize,
+        active: bool,
+    ) -> Healthchecker {
+        let start = tokio::time::Instant::now();
+        let now_fn = NowFn::from(move || start.elapsed().as_millis() as u64);
+
+        let storage_metadata = CollectionMetadata {
+            persist_location: (*PERSIST_LOCATION).clone(),
+            remap_shard: ShardId::new(),
+            data_shard: ShardId::new(),
+            status_shard: status_shard_id,
+        };
+
+        Healthchecker::new(
+            source_name,
+            upstream_name,
+            source_id,
+            source_type,
+            worker_id,
+            worker_count,
+            active,
+            &Arc::clone(&*PERSIST_CACHE),
+            &storage_metadata,
+            now_fn,
+        )
+        .await
+        .unwrap()
+    }
+
+    async fn simple_healthchecker(status_shard_id: ShardId, source_id: u64) -> Healthchecker {
+        new_healthchecker(
+            status_shard_id,
+            "source".to_string(),
+            GlobalId::User(source_id),
+            "kafka",
+            Some("sample-topic".to_string()),
+            1,
+            1,
+            true,
+        )
+        .await
+    }
+
+    async fn dump_storage_collection(shard_id: ShardId) -> Vec<Row> {
+        let persist_client = (*PERSIST_CACHE)
+            .lock()
+            .await
+            .open((*PERSIST_LOCATION).clone())
+            .await
+            .unwrap();
+
+        let (write_handle, read_handle) = persist_client.open(shard_id).await.unwrap();
+
+        let upper = write_handle.upper();
+        let readable_upper = Antichain::from_elem(upper.elements()[0] - 1);
+
+        let mut snapshot = read_handle.snapshot(readable_upper).await.unwrap();
+
+        let mut v: Vec<((Result<SourceData, String>, Result<(), String>), u64, i64)> = Vec::new();
+        while let Some(mut next) = snapshot.next().await {
```

Comment:
there is `read_all()` on `SnapshotIter` but it's behind `cfg(test)` and it seems that can't be used from tests of other modules. :see_no_evil:

Reply:
Yep, I did try using it, and this is an unashamed copy of that exactly because it didn't work :(

Reply:
it might work if you make that `cfg(any(test, debug_assertions))`?